### ISSUE TITLE
[Question] Parallel dynamic tree/graph search in Rayon

### ISSUE THREAD

**Issue by orenbenkiki** [rayon-rs/rayon#830]
What is the idiomatic way to do the following in Rayon:

* Starting at a given state
* A function that, given a state, generates additional states
* Generate all states in parallel, collecting them (e.g. into a vector).

The simplest way is to use an `Arc<RwLock<Vec<State>>>` and use a `ThreadPoolBuilder.install(|| reach_state(first_state))`, where `reach_state` will generate the sub-states, then `par_iter` on each generated sub-state, pushing it into the vector and recursively invoke `reach_state` on it.

This works. However, this consumes large amounts of stack stack due to the recursive calls to `reach_state`.

Is there an idiomatic way to setup a queue of items, and tell Rayon to process items in parallel from the queue, where this processing may add new items to the queue? This pattern is common in graph algorithms (basically, parallel variants of Dijkstra's algorithm, with either an explicit input graph or a graph that is being built dynamically as it is being searched).

The underlying libraries (e.g. crossbeam) can obviously do this, but with nothing like Rayon's ease of use.

**Comment by wagnerf42** [comment]
hi,

here is one:

```rust
use rayon::iter::plumbing::{bridge_unindexed, Folder, UnindexedConsumer, UnindexedProducer};
use rayon::prelude::*;
use std::collections::VecDeque;
use std::fmt::Debug;
use std::iter::once;
use std::marker::PhantomData;

struct TreeProducer<'b, S, B, I> {
    to_explore: VecDeque<Vec<S>>, // invariant: no one is empty here
    seen: Vec<S>,
    breed: &'b B,
    phantom: PhantomData<I>,
}

impl<'b, S, B, I> UnindexedProducer for TreeProducer<'b, S, B, I>
where
    S: Send + Debug,
    B: Fn(&S) -> I + Send + Sync,
    I: Iterator<Item = S> + Send,
{
    type Item = S;
    fn split(mut self) -> (Self, Option<Self>) {
        // explore while front is of size one.
        while self
            .to_explore
            .front()
            .map(|f| f.len() == 1)
            .unwrap_or(false)
        {
            let front_node = self.to_explore.pop_front().unwrap().pop().unwrap();
            self.to_explore
                .push_back((self.breed)(&front_node).collect());
            self.seen.push(front_node);
        }
        // now take half of the front.
        let f = self.to_explore.front_mut();
        let right_children = f.map(|f| {
            let n = f.len() / 2;
            let mut c = Vec::with_capacity(n);
            for _ in 0..n {
                c.push(f.pop().unwrap());
            }
            c
        });
        let right = right_children.map(|c| TreeProducer {
            to_explore: once(c).collect(),
            seen: Vec::new(),
            breed: self.breed,
            phantom: PhantomData,
        }); //TODO: if nothing we could split seen
        (self, right)
    }
    fn fold_with<F>(self, mut folder: F) -> F
    where
        F: Folder<Self::Item>,
    {
        // start by consuming everything seen
        for s in self.seen {
            folder = folder.consume(s);
            if folder.full() {
                return folder;
            }
        }
        // now do all remaining explorations
        for mut e in self.to_explore {
            while let Some(s) = e.pop() {
                e.extend((self.breed)(&s));
                folder = folder.consume(s);
                if folder.full() {
                    return folder;
                }
            }
        }
        folder
    }
}

struct TreeIter<S, B, I> {
    initial_state: S,
    breed: B,
    phantom: PhantomData<I>,
}

impl<S, B, I> ParallelIterator for TreeIter<S, B, I>
where
    S: Send + Debug,
    B: Fn(&S) -> I + Send + Sync,
    I: Iterator<Item = S> + Send,
{
    type Item = S;
    fn drive_unindexed<C>(self, consumer: C) -> C::Result
    where
        C: UnindexedConsumer<Self::Item>,
    {
        let producer = TreeProducer {
            to_explore: once(vec![self.initial_state]).collect(),
            seen: Vec::new(),
            breed: &self.breed,
            phantom: PhantomData,
        };
        bridge_unindexed(producer, consumer)
    }
}

fn descendants<S, B, I>(root: S, breed: B) -> TreeIter<S, B, I>
where
    S: Send,
    B: Fn(&S) -> I + Send + Sync,
    I: Iterator<Item = S> + Send,
{
    TreeIter {
        initial_state: root,
        breed,
        phantom: PhantomData,
    }
}

fn main() {
    let v: Vec<u32> = descendants(10, |e| {
        once(e / 2).chain(once(e / 2 + 1)).filter(|&e| e > 2)
    })
    .collect();
    eprintln!("v: {:?}", v);
}
```

note that you don't necessarily need to collect. if you do it also could be more efficient to fold into `Vec` and collect into `LinkedList<Vec>`

**Comment by orenbenkiki** [comment]
@wagnerf42 - thanks for the code. This got complex pretty fast... Especially compared to "replace `iter` with `par_iter`" :-)

I don't suppose there's some crate somewhere that has something similar to:

```rust
fn par_deque<T>(first: T, task: Fn(T, &RwLock<VecDeque<T>>);
```

So I could write something along the lines of:
```rust
par_deque(first_data, |data, pending| {
    ... do stuff with the data ...
    pending.write().unwrap().push_back(... more data ...);
    ... do stuff with the data ...
    pending.write().unwrap().push_back(... even more data ...);
    ... etc. ...
});
```

**Comment by cuviper** [comment]
The built-in way to do something like a `par_deque` is to use a `scope` and then `spawn` work into the queue. These can be nested, where one spawned job also adds further `spawn` calls, without being directly recursive. (Indirectly, they may still be layered on the stack due to work stealing.) That doesn't offer a way to `collect` though, so at the moment you would have to use your own synchronized means.

It would be interesting to think about something within the `ParallelIterator` API though, so you _could_ `collect`. The current `par_bridge` might be a decent starting point for this, being something that does manage its own queue, only it would be different in the way new stuff gets queued.

**Comment by wagnerf42** [comment]
> @wagnerf42 - thanks for the code. This got complex pretty fast... Especially compared to "replace `iter` with `par_iter`" :-)

well. the inner workings are not easy but the use is easy.

here is an example with a tree traversal.
(i fixed a bug btw)

If you look at the main it's just a call to `descendant` in order to get the parallel iterator.
the `descendant` function is literally what you wanted.

```rust
use rayon::iter::plumbing::{bridge_unindexed, Folder, UnindexedConsumer, UnindexedProducer};
use rayon::prelude::*;
use std::collections::VecDeque;
use std::fmt::Debug;
use std::iter::once;
use std::marker::PhantomData;

struct TreeProducer<'b, S, B, I> {
    to_explore: VecDeque<Vec<S>>, // invariant: no one is empty here
    seen: Vec<S>,
    breed: &'b B,
    phantom: PhantomData<I>,
}

impl<'b, S, B, I> UnindexedProducer for TreeProducer<'b, S, B, I>
where
    S: Send + Debug,
    B: Fn(&S) -> I + Send + Sync,
    I: Iterator<Item = S> + Send,
{
    type Item = S;
    fn split(mut self) -> (Self, Option<Self>) {
        // explore while front is of size one.
        while self
            .to_explore
            .front()
            .map(|f| f.len() == 1)
            .unwrap_or(false)
        {
            let front_node = self.to_explore.pop_front().unwrap().pop().unwrap();
            let next_to_explore: Vec<_> = (self.breed)(&front_node).collect();
            if !next_to_explore.is_empty() {
                self.to_explore.push_back(next_to_explore);
            }
            self.seen.push(front_node);
        }
        // now take half of the front.
        let f = self.to_explore.front_mut();
        let right_children = f.map(|f| {
            let n = f.len() / 2;
            let mut c = Vec::with_capacity(n);
            for _ in 0..n {
                c.push(f.pop().unwrap());
            }
            c
        });
        let right = right_children.map(|c| TreeProducer {
            to_explore: once(c).collect(),
            seen: Vec::new(),
            breed: self.breed,
            phantom: PhantomData,
        }); //TODO: if nothing we could split seen
        eprintln!("right: {}", right.is_some());
        (self, right)
    }
    fn fold_with<F>(self, mut folder: F) -> F
    where
        F: Folder<Self::Item>,
    {
        // start by consuming everything seen
        for s in self.seen {
            folder = folder.consume(s);
            if folder.full() {
                return folder;
            }
        }
        // now do all remaining explorations
        for mut e in self.to_explore {
            while let Some(s) = e.pop() {
                e.extend((self.breed)(&s));
                folder = folder.consume(s);
                if folder.full() {
                    return folder;
                }
            }
        }
        folder
    }
}

struct TreeIter<S, B, I> {
    initial_state: S,
    breed: B,
    phantom: PhantomData<I>,
}

impl<S, B, I> ParallelIterator for TreeIter<S, B, I>
where
    S: Send + Debug,
    B: Fn(&S) -> I + Send + Sync,
    I: Iterator<Item = S> + Send,
{
    type Item = S;
    fn drive_unindexed<C>(self, consumer: C) -> C::Result
    where
        C: UnindexedConsumer<Self::Item>,
    {
        let producer = TreeProducer {
            to_explore: once(vec![self.initial_state]).collect(),
            seen: Vec::new(),
            breed: &self.breed,
            phantom: PhantomData,
        };
        bridge_unindexed(producer, consumer)
    }
}

fn descendants<S, B, I>(root: S, breed: B) -> TreeIter<S, B, I>
where
    S: Send,
    B: Fn(&S) -> I + Send + Sync,
    I: Iterator<Item = S> + Send,
{
    TreeIter {
        initial_state: root,
        breed,
        phantom: PhantomData,
    }
}

#[derive(Debug)]
struct Node {
    content: u32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn main() {
    let t = Node {
        content: 10,
        left: Some(Box::new(Node {
            content: 3,
            left: None,
            right: None,
        })),
        right: Some(Box::new(Node {
            content: 14,
            left: None,
            right: Some(Box::new(Node {
                content: 18,
                left: None,
                right: None,
            })),
        })),
    };
    let v: Vec<&Node> = descendants(&t, |r| {
        r.left
            .as_ref()
            .into_iter()
            .chain(r.right.as_ref())
            .map(|n| &**n)
    })
    .collect();
    eprintln!("v: {:?}", v);
}
```

**Comment by orenbenkiki** [comment]
@cuviper - ah, scope seems to do almost exactly what I need (with a few lines of code). I'll give it a try. Thanks!

**Comment by orenbenkiki** [comment]
Was too quick to announce success.

I can make this work nicely using global functions. However I failed to make it work with method:

```rust
use core::marker::PhantomData;
use rayon::scope;
use rayon::Scope;
use rayon::ThreadPoolBuilder;

// Works
pub fn global_launch() {
    ThreadPoolBuilder::new().build().unwrap().install(|| {
        scope(|parallel_scope| {
            global_use(parallel_scope);
        })
    });
}

// Works
pub fn global_use(parallel_scope: &Scope) {
    parallel_scope.spawn(|same_scope| global_use(same_scope));
}

struct Plain;

impl Plain {
    pub fn plain_launch(&self) {
        ThreadPoolBuilder::new().build().unwrap().install(|| {
            scope(|parallel_scope| {
                self.plain_use(parallel_scope);
            })
        });
    }

    // Fails to compile
    pub fn plain_use(&self, parallel_scope: &Scope) {
        parallel_scope.spawn(|same_scope| self.plain_use(same_scope));
    }
}

struct Annotated<'a> {
    _phantom: PhantomData<&'a i32>,
}

impl<'a> Annotated<'a> {
    pub fn annotated_launch(&'a self) {
        ThreadPoolBuilder::new().build().unwrap().install(|| {
            scope(|parallel_scope| {
                self.a_is_b(parallel_scope);
                self.b_is_a(parallel_scope);
            })
        });
    }

    // Fails to compile
    pub fn a_is_b<'b>(&'a self, parallel_scope: &'b Scope<'b>)
    where
        'a: 'b,
    {
        parallel_scope.spawn(|same_scope| self.a_is_b(same_scope));
    }

    // Fails to compile
    pub fn b_is_a<'b>(&'a self, parallel_scope: &'b Scope<'b>)
    where
        'b: 'a,
    {
        parallel_scope.spawn(|same_scope| self.b_is_a(same_scope));
    }
}
```

I'm obviously missing something subtle about the lifetime annotations here. Is there a way to use method calls which spawn more tasks inside a parallel scope?

**Comment by cuviper** [comment]
Let's look at the signature for `Scope::spawn`:

```rust
    pub fn spawn<BODY>(&self, body: BODY)
    where
        BODY: FnOnce(&Scope<'scope>) + Send + 'scope,
```

The first thing (which isn't apparent) is that `Scope` is [invariant](https://doc.rust-lang.org/reference/subtyping.html#variance) in its lifetime `'scope`. This is because `Scope` is accepting arbitrary closures (like `BODY`) that we will guarantee to be completed before `'scope` ends. If we had variance that allowed it to act like `Scope<'smaller_scope>`, it would then implicitly accept closures with shorter lifetimes, but would still let them run until the actual end of `scope` -- potentially with expired borrows (dangling references). So the `'scope` must be fixed from the initial `scope(...)` call, which is where it waits for every `spawn` completion before returning.

The second thing is that when you see an anonymous lifetime in a `Fn` argument, this is effectively a [higher-ranked trait bound](https://doc.rust-lang.org/reference/trait-bounds.html#higher-ranked-trait-bounds) (HRTB), which roughly means the function must accept _any_ lifetime there. It's implicitly the same as `BODY: for<'any> FnOnce(&'any Scope<'scope>)`, and we only know `'scope: 'any` to be well-formed. (i.e. `'scope` outlives `'any`, because the object must be alive for its reference to be valid.) You're never given matching lifetimes like `&'a Scope<'a>`, and its invariance means a `Scope<'scope>` can't "shrink" to match `&'a` either.

So if we look at your first example:

```rust
    pub fn plain_use(&self, parallel_scope: &Scope) {
        parallel_scope.spawn(|same_scope| self.plain_use(same_scope));
    }
```

You omitted the lifetime in `Scope`, which might be ok, but your closure still has to outlive that to meet `BODY: 'scope`. It was fine in `global_use` because that closure didn't capture anything, but here you are capturing `&self` which has an unrelated lifetime, so the `BODY: 'scope` requirement cannot be met. Thus the error:

```text
   |
30 |     pub fn plain_use(&self, parallel_scope: &Scope) {
   |                      -----                   -----
   |                      |
   |                      these two types are declared with different lifetimes...
31 |         parallel_scope.spawn(|same_scope| self.plain_use(same_scope));
   |                        ^^^^^ ...but data from `self` flows into `parallel_scope` here
```

Furthermore, closures capture by reference when possible, so that captured type is actually an unwieldy `&&Self` tied to the local stack frame. You can avoid that with `move` to capture the direct type `&Self`. Here is the simplest way I would correct it:

```rust
    pub fn plain_use<'a>(&'a self, parallel_scope: &Scope<'a>) {
        parallel_scope.spawn(move |same_scope| self.plain_use(same_scope));
    }
```

Here is a very explicit form with lifetimes on everything:

```rust
    pub fn plain_use<'ref_self, 'ref_scope, 'scope>(
        &'ref_self self,
        parallel_scope: &'ref_scope Scope<'scope>,
    ) where
        'ref_self: 'scope,
        'scope: 'ref_scope,
    {
        parallel_scope.spawn(move |same_scope| self.plain_use(same_scope));
    }
```

I hope that's enough for you to figure out what's wrong with your `Annotated`, but feel free to ask further questions!

**Comment by orenbenkiki** [comment]
Wow, thanks for the great and detailed reply!

I understand the need to ensure the lifetimes nest properly, but I am obviously still clueless when it comes to comminicating it to the compiler. The explicit form was a great help. The top about the move is also great - I haven't considered that it would use &&Self. My intuition about these things still has a long way to go.

At any rate, with the additional annotations along the lines you described everything works perfectly now. Thanks again!
